### ISSUE TITLE
how to lint "must give optional props default values" in typescript

### ISSUE THREAD

**Issue by XHMM** [yannickcr/eslint-plugin-react#2317]
For example, is there a rule to lint **below is wrong**:
```js
const MyInput: FunctionComponent<IProps> = ({ label, type }) => {
  return (
    <div>
      <label>{label}</label>
      <input type={type} />
    </div>
  );
};
interface IProps {
  label: string;
  type?: string; // it's an optional prop
}
```
but **below is right**:
```js
const MyInput: FunctionComponent<IProps> = ({ label, type='text' }) => { 
 // ... same as above
}
```

I known there is a `react/require-default-props` rule , but it's not suitable for `typescipt`.

**Comment by ljharb** [comment]
It’s not wrong; the issue is that in both flow and typescript, there are actually *two* props signatures for a component: the external one (propTypes) and the internal one (propTypes + defaultProps).

If you’re not defining both, and you have any optional props, your types are wrong.

**Comment by ljharb** [comment]
Separately, no, it’s not right to use default arguments. Using .defaultProps is the proper method.

**Comment by XHMM** [comment]
Thanks for reply. 

So do you mean I must use PropTypes definition to let eslint work for my purpose？ 

And no other rules or plugins can do this without writing PropType？

**Comment by ljharb** [comment]
You should always write a runtime `.propTypes` imo - but eslint should be able to work as long as your external *and* internal props are typed somehow.

**Comment by JustFly1984** [comment]
Fwy according to Den Abramov, defaultProps will be deprecated

**Comment by ljharb** [comment]
a) cite your source; b) dan is not the emperor of react; c) the best practice may change when react actually deprecates these things in a non-alpha release, but not a moment before.

**Comment by XHMM** [comment]
Hope there is a specific  rule for typescript component without writing PropTypes. 👐  

Closed because don't know what to discuss.

**Comment by tmkasun** [comment]
If someone gets into the same problem, here are some of the sources that discuss deprecating `defaultProps`

- [Dan's tweet](https://twitter.com/dan_abramov/status/1133878326358171650)
- [RFC](https://github.com/reactjs/rfcs/pull/107)

**Comment by ljharb** [comment]
It’s approaching 3 years since that RFC was opened, and no movement yet. It would be a very poor decision for that rfc to land, so it’s best to wait until there’s actual movement there.
